PEP 788: Address feedback from first discussion round

ZeroIntensity opened the pull request:
* Change is either:
    * [x] To a Draft PEP
    * [ ] To an Accepted or Final PEP, with Steering Council approval
    * [ ] To fix an editorial issue (markup, typo, link, header, etc)
* [x] PR title prefixed with PEP number (e.g. ``PEP 123: Summary of changes``)

ZeroIntensity commented:
@godlygeek I can't add you as a reviewer, but I'd appreciate your opinion on some of the new things I'm proposing here.

ZeroIntensity commented:
Bump @AA-Turner @vstinner now that the beta is out. There's no rush considering we have a year until the 3.15 freeze, though.

ZeroIntensity commented on peps/pep-0788.rst:
Yeah, I'm imagining `PyInterpreterRef` as an alias to `uintptr_t`, and `0` is the equivalent to `NULL`. I'd rather make it a pointer instead of an API like that.

ZeroIntensity commented on peps/pep-0788.rst:
That's probably harder to use, though.

vstinner replied:
I expect PyInterpreterRef and PyInterpreterWeakRef to be handle-like: not numbers. For example, `typedef struct { uintptr_t _bits; } PyInterpreterRef;`.

I would prefer to avoid creating "NULL" PyInterpreterRef. It avoids having to think about them in each function: should it be accepted? what's the behavior?

ZeroIntensity commented on peps/pep-0788.rst:
```diff
@@ -649,46 +871,34 @@ they can still be used with this API:
-        if (interp == NULL) {
-            fputs("Python has shut down", stderr);
+        PyInterpreterWeakRef *wref = (PyInterpreterWeakRef *)arg;
+        PyInterpreterRef *ref = PyInterpreterWeakRef_AsStrong(wref);
```
Oops.

vstinner commented:
I'm not comfortable with `PyInterpreterRef` and `PyInterpreterWeakRef` which are different: `PyInterpreterRef` is used as a scalar value, whereas `PyInterpreterWeakRef` is used as a pointer. I would prefer that both are scalars, or both are pointers. Maybe using a pointer API would be more future proof and it makes sure that the value (the pointer) can be stored ... in a pointer, for a callback argument.

ZeroIntensity commented on peps/pep-0788.rst:
> You might rename the parameter to just "ref". Same remarks for PyInterpreterWeakRef_AsStrong().

Why? Isn't that just less clear?

vstinner replied:
To not associate NULL value (`void*`) with PyInterpreterRef.

ZeroIntensity replied:
Right, that's why it's named `ref_ptr` and not `ref`. Pointers can always be `NULL`, but interpreter references can't--don't we lose that distinction by naming the parameter `ref`?

vstinner replied:
Setting `ref_ptr` to NULL would be useless. I suppose that you mean setting `*ref_ptr` to NULL here. My point is that it should not be possible to compare a reference to NULL, or to create a reference without the API.

ZeroIntensity replied:
Oh right, I forgot it's not a pointer-to-pointer. I'll remove that part.

colesbury commented on peps/pep-0788.rst:
```diff
@@ -131,18 +189,26 @@ the thread:
+its thread state to attach, while thread B holds the GIL and is waiting on the
+lock.
+
+On free-threaded builds, lock-ordering deadlocks are still possible
+if thread A acquired the lock for object A and then object B, and then
+another thread tried to acquire those locks in the reverse order. Free-threading
+currently protects against this by releasing locks when the thread state is
+detached, making detachment a necessity to prevent deadlocks.
```
I think this is mixing up a few ideas (critical sections and stop the world). You want `Py_BEGIN_ALLOW_THREADS` around blocking operations to prevent deadlock with GC (or other stop-the-world) events. 

Stop-the-world events are analogous to holding the GIL -- one thread has exclusive access to the interpreter -- so you can end up with the same deadlocks.

ZeroIntensity replied:
Hm, is the current thing I wrote wrong? I think this paragraph is quite long already, and it's only here to quickly explain why `Py_BEGIN_ALLOW_THREADS` is still needed on free-threaded builds. If there's nothing wrong with what's already there, it's probably fine to leave it.

colesbury replied:
I agree that it's getting long already. Was this expanded in response to feedback on the PEP? I would just have one sentence regarding the free threaded build at the end of the lock ordering deadlock paragraph. Something like:

...while thread B holds the GIL and is waiting on the lock. A similar deadlock can occur in the free threaded build during stop the world pauses, which happen during during garbage collection.

colesbury replied:
> Hm, is the current thing I wrote wrong?

The first part is definitely true: "on free-threaded builds, lock-ordering deadlocks are still possible". 

The subsequent explanation doesn't make much sense to me. The "releasing locks when the thread state is
detached" behavior is only applicable to critical sections, and we don't use `Py_BEGIN_ALLOW_THREADS` directly with critical section APIs. The critical section API implementations do the detaching/reattaching internally.

ZeroIntensity replied:
> Was this expanded in response to feedback on the PEP? I would just have one sentence regarding the free threaded build at the end of the lock ordering deadlock paragraph.

Ok, yeah, I'll do it this way.

> The critical section API implementations do the detaching/reattaching internally.

I was thinking of something like this:

```c
Py_BEGIN_CRITICAL_SECTION(whatever);
acquire_os_lock(); // NOT PyMutex
Py_END_CRITICAL_SECTION();
```

You still need to have `Py_BEGIN_ALLOW_THREADS` to release that critical section. Otherwise you can get lock-ordering deadlocks, right?

ZeroIntensity replied:
Section is now removed.

colesbury commented on peps/pep-0788.rst:
```diff
@@ -131,18 +189,26 @@ the thread:
-Daemon threads can cause finalization deadlocks
-***********************************************
+Daemon Threads Can Deadlock Finalization
+****************************************
```
The scenario here that leads to deadlock here happens when you acquire a lock during an object destructor/finalizer thats also acquired elsewhere. But if you're doing that, non-daemon threads won't save you because you will still occasionally deadlock with the GC.

ZeroIntensity replied:
Hmm, I wasn't aware of this. That sounds like it compromises some of the motivation behind the PEP. Could you go into a little more detail here?

colesbury replied:
Python's GC runs object finalizers (i.e., `__del__` methods) in the same thread that triggered the GC, so you can end trying to reacquire a lock you already hold.

Here's an example from stackoverflow: https://stackoverflow.com/questions/18774401/self-deadlock-due-to-garbage-collector-in-single-threaded-code

Java runs finalizers in their own thread to avoid this sort of problem.

Generally, the "fix" in Python is to avoid acquiring locks in `__del__` functions.

ZeroIntensity replied:
This section is now removed.

colesbury commented on peps/pep-0788.rst:
```diff
@@ -189,30 +254,28 @@ error, as noted in `python/cpython#124622 <https://github.com/python/cpython/iss
 
 At the time of writing, the current behavior of :c:func:`PyGILState_Ensure` does not
 match the documentation. Instead of hanging the thread during finalization
 as previously noted, it's extremely common for it to crash with a segmentation
```
Is this "extremely common"? The link is to a core developer bug report with no comments from users?

ZeroIntensity replied:
It is, but there's not one issue to sum it up; Eric's issue is simply a tracker. I've encountered a couple of reports before regarding `PyGILState_Ensure` crashes, and it was also [noted](https://discuss.python.org/t/safely-using-the-c-api-when-python-might-shut-down/78850) in the `PyGILState_EnsureOrSafelyFail` thread:

> Sadly, there isn’t a reliable way to do this in C++, and it usually segfaults the application

I'll reword it, though. I think "extremely common" is confusing.

colesbury commented on peps/pep-0788.rst:
```diff
@@ -701,48 +916,90 @@ For example:
+Example: Calling Python Without a Closure
+*****************************************
+
+There are a few cases where callback functions don't take a closure
+(``void *arg``), so it's impossible to acquire a reference to any specific
```
A closure is a function that captures its environment. The combined callback and argument behave kind of like a closure, but calling the `void *arg` by itself a "closure" seems strange to me. I'd call it a "callback parameter" or "callback argument"

colesbury commented on peps/pep-0788.rst:
```diff
@@ -701,48 +916,90 @@ For example:
+
+But wait, won't that break with subinterpreters, per
+:ref:`pep-788-subinterpreters-gilstate`? Fortunately, since the callback has
+no closure, it's not possible for the caller to pass any objects or
+interpreter-specific data, so it's completely safe to choose the main
```
I don't see how "it's completely safe to choose the main interpreter" follows from "the callback has no closure".

Maybe these use cases (like `readline`) only make sense in the main interpreter, but choosing the correct interpreter seems important.

ZeroIntensity replied:
If you don't have a closure argument/callback parameter or some other workaround, then you can't pass interpreter-specific data, so you're good to assume that the main interpreter is the one you want. It might not be the right choice, but by "totally safe", I mean that things will just raise a completely safe exception (e.g. an `AttributeError`), rather than crashing through using an object from a different interpreter.

So basically, if you can access state from the caller, then you can also store the interpreter reference, and if not, then there's no way to access invalid cross-interpreter state anyway.

colesbury reviewed (commented):
When does `PyThreadState_Ensure()` create a new thread state instead of reusing an existing one?

ZeroIntensity commented on peps/pep-0788.rst:
```diff
@@ -14,85 +14,106 @@ Post-History: `10-Mar-2025 <https://discuss.python.org/t/83959>`__,
+  deallocated, and shutdown for the main interpreter includes the entire Python
+  runtime being finalized.
+
+Native and Python Threads
```
I don't totally disagree, but I'd really prefer not using the clunky phrase "non-Python created thread", especially considering how many times "native thread" is used throughout the PEP. Is there a shorter term we could use?

vstinner replied:
OS threads?

ZeroIntensity replied:
I think "OS thread" has the same problem as "native thread". I'm personally fine with still using the term "native thread", because this PEP will change how `threading` works too. So you get two interpretations, both being correct:

- "Reimagining Threads *natively*": The PEP changes how to use OS threads with a native caller.
- "Reimagining *Native* Threads": The PEP changes how native OS threads, `threading` included, interact with the interpreter.

vstinner replied:
If you want to avoid "native threads", just write "non-Python threads"?

ZeroIntensity commented on peps/pep-0788.rst:
```diff
@@ -14,85 +14,106 @@ Post-History: `10-Mar-2025 <https://discuss.python.org/t/83959>`__,
+called. There's a subtle difference between the two terms, as used in this
+PEP:
+
+- "Finalization" refers to an interpreter getting ready to "shut down", in
```
It's not important for the main interpreter, but there's an important distinction for subinterpreters because the `PyInterpreterState` structure becomes invalid after *shutdown*, but not during finalization. That's an important detail for the rationale on why we're not using `PyInterpreterState *` for `PyThreadState_Ensure`.

But I think you're right, anyway. I'll remove this.

ZeroIntensity replied:
Removed.

ZeroIntensity commented:
> When does PyThreadState_Ensure() create a new thread state instead of reusing an existing one?

The same places where `PyGILState_Ensure` would, in addition to when the current interpreter doesn't match the requested interpreter. Should that be specified, or left up the implementation? I could see it going either way.

ZeroIntensity commented:
Here's what I was envisioning:

```c
// No tstate ever in this thread
// current: NULL

// New thread state for the main interpreter
PyThreadState_Ensure(main_interp);
// current: main

// Do nothing
PyThreadState_Ensure(main_interp);
// current: main

Py_BEGIN_ALLOW_THREADS
// current: NULL

// Reattach the old thread state
PyThreadState_Ensure(main_interp);

// current: main

// Detach it again
PyThreadState_Release();
// current: NULL

Py_END_ALLOW_THREADS
// current: main

// New thread state for the subinterpreter
PyThreadState_Ensure(subinterp);
// current: subinterp

// New thread state for the main interpreter
PyThreadState_Ensure(main_interp);
// current: main
```

Basically, `PyThreadState_Ensure` for a different interpreter treats it as a new thread.

vstinner commented:
cc @pablogsal: the PEP update I told you about. You might want to have a look.

vstinner commented on peps/pep-0788.rst:
I suggest to rename it to "Deprecating ..." since the API will stay for many years.

ericsnowcurrently commented on peps/pep-0788.rst:
```diff
@@ -14,85 +14,106 @@ Post-History: `10-Mar-2025 <https://discuss.python.org/t/83959>`__,
 
-Native threads will always hang during finalization
----------------------------------------------------
+Native Threads Always Hang During Finalization
```
Can we address this with a distinct return value and two new functions?

* `int PyGILState_Failed(PyGILState_STATE)`
* `void PyGILState_SetHangDuringFini(int)`  # default 1

ericsnowcurrently replied:
Another option is to support a callback that gets called when the interpreter is finalizing:

```c
typedef int (tstate_fini_callback *)(PyThreadState *, void *arg);
void PyThreadState_SetFiniCallback(PyThreadState *, tstate_fini_callback, void *arg);
```

I'm sure there are other options that are likewise compatible with the existing API (not that a replacement for `PyGILState_Ensure()` is necessarily a bad idea).

ericsnowcurrently replied:
FWIW, at a certain point it make more sense to simply replace the old API, as you are proposing, rather than add a bunch of functions to awkwardly work around the deficiencies of the original API.  That very well may be the case here.  I just want to be sure we consider the obvious alternatives.

ZeroIntensity replied:
That sounds sort of racy. `PyGILState_Ensure` will always have to hang the thread by default, I don't think we can effectively change that without breaking code. Since that's the default, the initial `PyGILState_Ensure` might hang the thread before you get a chance to call something like `PyGILState_SetHangDuringFini`.

> int PyGILState_Failed(PyGILState_STATE)

How would this work with old code? I'm normally not one to quote the zen, but I think this is definitely a case where "errors should never pass silently."

---

> FWIW, at a certain point it make more sense to simply replace the old API, as you are proposing, rather than add a bunch of functions to awkwardly work around the deficiencies of the original API. That very well may be the case here. I just want to be sure we consider the obvious alternatives.

Yeah, I've gone through several different iterations of the overall concept. There was a prior effort for `PyGILState_EnsureOrSafelyFail`, if you're interested in some of the history behind this proposal. But, I think we're better off just starting from scratch. I'm not confident that we'll be able to retrofit the existing APIs with something that works well.

ZeroIntensity replied:
I'm going to resolve this so we can merge. Feel free to bring this up on the DPO thread.

ericsnowcurrently commented on peps/pep-0788.rst:
```diff
@@ -101,23 +122,60 @@ For example, a callback that wants to call Python code might be invoked when:
-*addition* to calling Python).
+code in their stream of calls.
+
+Joining the Thread isn't Always Possible
```
How is this different from the previous one?

ZeroIntensity replied:
Hm, what do you mean?

ericsnowcurrently replied:
The previous motivation section, "Native Threads Always Hang During Finalization".

ZeroIntensity replied:
This section is about using something like an `atexit` handler to join threads before finalization as a *solution* to hanging, while the previous section is about the hanging itself. So, it's like a disease vs. cure situation here.

ericsnowcurrently commented on peps/pep-0788.rst:
```diff
@@ -101,23 +122,60 @@ For example, a callback that wants to call Python code might be invoked when:
+it). If the interpreter is finalizing or has shut down, then the thread is
+hung, disrupting the C++ caller.
+
+``Py_IsFinalizing`` is Insufficient
```
Fixing the race here is certainly worth doing, but seems like a separate issue from the GILState API.  It also doesn't seem like something that needs a PEP.

ZeroIntensity replied:
It's not a race; it's a time-of-check to time-of-use problem. `Py_IsFinalizing` couldn't ever reliably be used as a guard against finalization because the interpreter could always start finalizing right after it has been called.

ericsnowcurrently commented on peps/pep-0788.rst:
```diff
@@ -131,18 +189,26 @@ the thread:
 
-Daemon threads can cause finalization deadlocks
-***********************************************
+Daemon Threads Can Deadlock Finalization
```
How much is this an issue in practice?

ZeroIntensity replied:
I'm not sure. I'm up for removing this section, because Sam already made a good point that locking isn't safe in finalizers, regardless of daemon threads, because of the garbage collector.

ZeroIntensity replied:
This section is now removed.

ericsnowcurrently commented on peps/pep-0788.rst:
```diff
@@ -189,30 +254,28 @@ error, as noted in `python/cpython#124622 <https://github.com/python/cpython/iss
+:c:func:`PyGILState_Ensure`, is `the most common <https://grep.app/search?q=pygilstate_ensure>`_.
 
-``PyGILState_Ensure`` generally crashes during finalization
+``PyGILState_Ensure`` Generally Crashes During Finalization
```
Crashes here are bugs to be fixed, which I don't expect will need a PEP to solve.

ZeroIntensity replied:
See my previous comment.

ericsnowcurrently commented on peps/pep-0788.rst:
```diff
@@ -189,30 +254,28 @@ error, as noted in `python/cpython#124622 <https://github.com/python/cpython/iss
 the existing crashes caused by :c:func:`PyGILState_Ensure`.
 
-The term "GIL" is tricky for free-threading
+The Term "GIL" is Tricky for Free-threading
```
Sure, the name can be a little confusing.  That is more of a "while we're at it, let's use a better name" situation than a motivation for a PEP.  I wouldn't make much of a point about it, except maybe to mention it in passing.

ZeroIntensity replied:
I think this is actually more important than it might seem. I've seen something along the lines of "we don't need `PyGILState_Ensure` on free-threading, right?" come up several, several times, including from core devs!

ZeroIntensity replied:
I'm going to resolve this so we can merge. Feel free to bring this up on the DPO thread.

ericsnowcurrently reviewed (commented):
Thanks for working on this.  I haven't gotten though the whole proposal, but did start by trying to understand what we're trying to solve here.  I've left short notes for each of the motivation sections.  Ultimately, I think are some bugs to be addressed separately and the other concerns could be solved without replacing the existing API.  That doesn't mean a new API isn't warranted, but let's see how much value it would add first.  (I'd  be glad to hop into a call to discuss this more.)

ZeroIntensity commented on peps/pep-0788.rst:
```diff
@@ -189,30 +254,28 @@ error, as noted in `python/cpython#124622 <https://github.com/python/cpython/iss
 
-The existing APIs are broken and misleading
--------------------------------------------
+The GIL-state APIs are Buggy and Confusing
```
The "confusing" part alludes to the "GIL" in the name, which has been a pain for free-threading. The bugginess is also important because the ideal way to fix the bugs is to implement this PEP :). (Basically, preventing finalization in critical areas like `PyGILState_Ensure` is the ideal way to fix it, which is what this PEP specifies!)

ZeroIntensity commented on peps/pep-0788.rst:
```diff
@@ -224,254 +287,290 @@ created by the authors of this PEP:
------------------------------------------------------
+.. _pep-788-subinterpreters-gilstate:
+
+``PyGILState_Ensure`` Doesn't Guess the Correct Interpreter
```
Eh... that seems messy, and doesn't solve the issue where an interpreter is short-lived if we're using a `PyInterpreterState *`. I really think we should just do away with `PyGILState` and *reimagine* how we do this.

ZeroIntensity replied:
I'm going to resolve this so we can merge. Feel free to bring this up on the DPO thread.

ZeroIntensity commented:
> I'd be glad to hop into a call to discuss this more.

That'd be great. Shoot me a message on Discord so we can figure something out. (Note that I'll have limited availability this week.)

ZeroIntensity commented on peps/pep-0788.rst:
```diff
@@ -14,85 +14,106 @@ Post-History: `10-Mar-2025 <https://discuss.python.org/t/83959>`__,
-like this:
-
-.. code-block:: c
+In Python, threads are able to interact with an interpreter (e.g., invoke the
```
I redid the abstract. Let me know if it's any better.

ZeroIntensity commented:
I think most of Eric's comments will fit better on the public discussion thread, so I didn't deal with them too much.

@colesbury I've addressed most of your concerns. In particular, thread states and interpreter refs are now separate, so there's no more "non-daemon thread states." I'd appreciate your rubber-stamp before landing this, would you mind doing a quick once-over when you get a chance?

vstinner commented on peps/pep-0788.rst:
```diff
@@ -14,85 +14,77 @@ Post-History: `10-Mar-2025 <https://discuss.python.org/t/83959>`__,
-           the reference count of the interpreter. */
-        Py_RETURN_NONE;
-    }
+Native threads are typically created by :c:func:`PyGILState_Ensure`, but more
```
PyGILState_Ensure() doesn't create a thread. It only attachs a Python thread state to an existing thread. The sentence is misleading.

vstinner commented on peps/pep-0788.rst:
```diff
@@ -101,23 +93,41 @@ For example, a callback that wants to call Python code might be invoked when:
+(to yield the :term:`GIL` to a different thread), or when a C function exits a
+:c:macro:`Py_BEGIN_ALLOW_THREADS` block, so simply guarding against whether the
+interpreter is finalizing isn't enough to safely call Python code. (Note that hanging
+the thread is relatively new behavior; in prior versions, the thread would terminate,
```
```suggestion
the thread is relatively new behavior; in prior versions, the thread would exit,
```

vstinner commented on peps/pep-0788.rst:
"ote that": is it a typo or "note that"?

vstinner commented on peps/pep-0788.rst:
I don't think that it's still useful, I suggest removing it. It can be added later if needed.

ZeroIntensity replied:
Same here, let's keep it just for completeness.

vstinner replied:
I don't think that completeness is a reason good enough to add a function. It needs an use case. Since PyThreadState_Ensure() doesn't "consume" a strong reference anymore, I don't see the point to duplicate a strong reference. The function looks useless to me. Same remark for PyInterpreterWeakRef_Dup().

vstinner replied:
In private, Peter explains to me that a thread spawning a thread may require a strong reference and it can be messy if the two threads share the same strong reference: it's unclear which one should close it and how. It's better that that each thread has its own strong reference and so PyInterpreterRef_Dup() is useful. I'm not fully convinced, but I don't want to hold the PEP update for that, so I close my request.

vstinner commented on peps/pep-0788.rst:
I don't think that it's still useful, so I suggest removing it. It can be added later if needed.

ZeroIntensity replied:
I'd rather keep it here for completeness.

vstinner commented on peps/pep-0788.rst:
I dislike this function since it's documented as "it can crash" :-( Do we really need this API? The https://pep-previews--4400.org.readthedocs.build/pep-0788/#example-calling-python-without-a-callback-parameter example uses it with PyInterpreterState_Main(), but PyInterpreterState_Main() is wrong if the program uses sub-interpreters.

ZeroIntensity replied:
Yeah, this function is a weird case that I'm not sure what to do about. It exists as a fallback for people who want to keep incompatibility with subinterpreters, but don't want to be broken by the removal of `PyGILState_Ensure`. The main interpreter is statically allocated, so it's the only interpreter that can't cause this function to crash.

Maybe we want something like `PyInterpreterRef_Main` instead?

vstinner replied:
I would prefer a PyInterpreterRef_GetMain() function if it allows removing PyInterpreterState_AsStrong().

ZeroIntensity replied:
Done.

vstinner commented on peps/pep-0788.rst:
I expected: "Return -1 on error". I don't understand what "non-zero with the old attached thread state restored" means.

vstinner commented on peps/pep-0788.rst:
```diff
@@ -630,12 +772,12 @@ they can still be used with this API:
         }
-        (void)PyThreadState_SetDaemon(1);
+        PyInterpreterRef_Close(ref);
         if (PyRun_SimpleString("print(42)") < 0) {
```
I don't understand well this example. Is it possible that the thread hangs instead of printing 42? It would help to explain the behavior.

ZeroIntensity replied:
Yes, printing will detach the thread state to write to stdout, and then can hang upon reattachment.

vstinner commented on peps/pep-0788.rst:
```diff
@@ -649,100 +791,171 @@ they can still be used with this API:
-        if (interp == NULL) {
-            fputs("Python has shut down", stderr);
+        ThreadData *data = (ThreadData *)arg;
+        PyInterpreterWeakRef wref = data->wref;
```
I would expect `PyMem_Free(data)` here.

ZeroIntensity replied:
It's technically valid to just let it leak. The callback will also leak here. I don't think cleaning up resources is particularly important for this example.

ZeroIntensity commented on peps/pep-0788.rst:
```diff
@@ -224,254 +287,290 @@ created by the authors of this PEP:
 
-Interpreters can concurrently shut down
-***************************************
+Concurrent Interpreter Deallocation is Frustrating
```
I'm going to resolve this so we can merge. Feel free to bring this up on the DPO thread.

ZeroIntensity commented:
I had to resolve many of the comments because it won't let us merge this while things are open. If there's anything that wasn't fully addressed, please bring it up on the thread!

@vstinner, we're ready to go.

vstinner commented:
I merged the PR, thanks.
